### PR TITLE
Make it easier to register configuration extension ...

### LINKED ISSUE #9529: Make it easier to register configuration extension options
Also, inspired by simplification you've done in `SessionContext`, maybe we should add companion object to `FunctionFactory` so it can register configuration as well. At the moment its a bit of work to register configuration:

https://github.com/milenkovicm/torchfusion/blob/634c2a4e39b81e1969671db8faea50b96bc43c7f/src/lib.rs#L101

```rust
let mut session_config = SessionConfig::new().with_information_schema(true);

session_config
      .options_mut()
      .extensions
      // register torch factory configuration
      .insert(TorchConfig::default());
```

it would be much better if config could be registered when `ctx.with_function_factory(...)` called ... but that could be a follow up PR

_Originally posted by @milenkovicm in https://github.com/apache/arrow-datafusion/issues/9482#issuecomment-1986983413_

### PR DESCRIPTION
**Pull request by milenkovicm**
... options

closes #9529

## Which issue does this PR close?

Closes #9529.

### CODE REVIEW

File: datafusion/execution/src/config.rs [hunk 1]
```diff
@@ -198,6 +201,12 @@ impl SessionConfig {
         self
     }
 
+    /// Insert new [ConfigExtension]
+    pub fn with_option_extension<T: ConfigExtension>(mut self, extension: T) -> Self {
```
Comment (milenkovicm) [inline datafusion/execution/src/config.rs:205]: we have discussed `with_extension_options` as the method name, but it make more sense to me to name it like this, not to be confused with `with_extension`

Comment (alamb) [reply datafusion/execution/src/config.rs:205]: Yeah I agree this is a reasonable name, even though it still a bit awkward

**Review by alamb — APPROVED** [review]
Thanks @milenkovicm

File: datafusion/execution/src/config.rs (hunk 1 above)
Comment (milenkovicm) [reply datafusion/execution/src/config.rs:205]: I'm horrible with naming things 😀 open to suggestions

Comment (alamb) [reply datafusion/execution/src/config.rs:205]: Yeah, if I had a better alternative I would give it.

**Comment by alamb** [comment]
Thanks again @milenkovicm

**Pull request merged**